### PR TITLE
feat: popup wallet entry

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-4328

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/popups/pages/Wallet/SetPaymentPassword/index.tsx [hunk 1]
```diff
@@ -92,64 +133,124 @@ const SetPaymentPassword = memo(() => {
 
     const onSubmit = handleSubmit(onConfirm)
 
-    useTitle(t('popups_set_the_payment_password'))
+    console.log({ wallets })
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/SetPaymentPassword/index.tsx]:
```suggestion
```

File: packages/mask/src/extension/popups/pages/Wallet/components/StartUp/index.tsx [hunk 2]
```diff
@@ -1,135 +1,127 @@
...
+        fontWeight: 700,
     },
-    content: {
-        flex: 1,
-        backgroundColor: '#f7f9fa',
-        paddingBottom: 40,
+    description: {
+        color: theme.palette.maskColor.third,
+        fontSize: 12,
+        fontFamily: 'Helvetica',
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/components/StartUp/index.tsx]: Use the inherited font family.

File: packages/dashboard/src/pages/CreateMaskWallet/AddDeriveWallet/index.tsx [hunk 3]
```diff
@@ -83,7 +82,7 @@ const AddDeriveWallet = memo(function AddDeriveWallet() {
 
     const { loading, value: dataSource } = useAsync(async () => {
         if (mnemonic) {
-            const unDeriveWallets = Array.from(indexes)
+            const unDeriveWallets = Array.from(indexes.current)
```
Comment (UncleBill) [inline packages/dashboard/src/pages/CreateMaskWallet/AddDeriveWallet/index.tsx:89]: Do not mix selected status into `dataSource`, maintain a `selectedAddresses` list and pass to the table instead. Thinking of a multiple `<select />`

Comment (hanszhoou) [reply packages/dashboard/src/pages/CreateMaskWallet/AddDeriveWallet/index.tsx:89]: We need `pathIndex` to generate wallet.

```ts
PluginServices.Wallet.recoverWalletFromMnemonic(
    `${walletName}${pathIndex}`,
    mnemonic,
    `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/${pathIndex}`,
),
```

Comment (UncleBill) [reply packages/dashboard/src/pages/CreateMaskWallet/AddDeriveWallet/index.tsx:89]: It's possible to get the index by `array.indexOf(address)`?

Comment (hanszhoou) [reply packages/dashboard/src/pages/CreateMaskWallet/AddDeriveWallet/index.tsx:89]: `pathIndex !== index`

File: packages/web3-providers/src/Web3/EVM/providers/MaskWallet.ts [hunk 4]
```diff
@@ -116,12 +116,14 @@ export class MaskWalletProvider
     }
 
     override async addWallet(wallet: Wallet): Promise<void> {
+        if (!this.hostedAccount && !this.wallets.length) this.walletStorage?.account.setValue(wallet.address)
```
Comment (guanbinrui) [inline packages/web3-providers/src/Web3/EVM/providers/MaskWallet.ts]:
```suggestion
        if (!this.hostedAccount && !this.wallets.length) await this.walletStorage?.account.setValue(wallet.address)
```

File: packages/web3-providers/src/Web3/EVM/providers/MaskWallet.ts [hunk 5]
```diff
@@ -116,12 +116,14 @@ export class MaskWalletProvider
...
         await this.context?.addWallet(wallet.address, wallet)
     }
 
     override async removeWallet(address: string, password?: string | undefined): Promise<void> {
         const scWallets = this.wallets.filter((x) => isSameAddress(x.owner, address))
         if (scWallets.length) await super.removeWallets(scWallets)
+        if (isSameAddress(this.hostedAccount, address)) this.walletStorage?.account.setValue('')
```
Comment (guanbinrui) [inline packages/web3-providers/src/Web3/EVM/providers/MaskWallet.ts]:
```suggestion
        if (isSameAddress(this.hostedAccount, address)) await this.walletStorage?.account.setValue('')
```

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 6]
```diff
@@ -162,14 +163,8 @@ export async function renameWallet(address: string, name: string) {
...
-export async function removeWallet(address: string, unverifiedPassword: string) {
-    await password.verifyPasswordRequired(unverifiedPassword)
-
-    // delete a wallet with derivationPath is not allowed
-    const wallet = await database.getWalletRequired(address)
-    if (wallet.derivationPath) throw new Error('Illegal operation.')
-
-    await database.deleteWallet(wallet.address)
+export async function removeWallet(address: string, unverifiedPassword?: string) {
+    await database.deleteWallet(address)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts]: For a dangerous action, must verify the password.

File: packages/web3-providers/src/Web3/EVM/providers/BaseHosted.ts [hunk 7]
```diff
@@ -22,7 +22,7 @@ export class BaseHostedProvider
     extends BaseProvider
     implements WalletAPI.Provider<ChainId, ProviderType, Web3Provider, Web3>
 {
-    private walletStorage:
+    public walletStorage:
```
Comment (guanbinrui) [inline packages/web3-providers/src/Web3/EVM/providers/BaseHosted.ts]: Don't access it outside. Please keep it a single source of mutations.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts [hunk 8]
```diff
@@ -235,20 +230,19 @@ export async function recoverWalletFromMnemonic(
     name: string,
     mnemonic: string,
     derivationPath = `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/0`,
-    initialPassword?: string,
+    password?: string,
 ) {
-    const password_ = initialPassword ?? (await password.INTERNAL_getPasswordRequired())
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/index.ts:240]: Try to get the password in-memory if the user doesn't provide it.

File: packages/mask/src/extension/popups/pages/Wallet/index.tsx [hunk 9]
```diff
@@ -101,16 +103,22 @@ export default function Wallet() {
...
     useEffect(() => {
         return CrossIsolationMessages.events.requestsUpdated.on(({ hasRequest }) => {
             if (hasRequest) retry()
         })
     }, [retry])
 
+    useEffect(() => {
+        if (!wallet || loadingHasPassword || isLocked || getLockStatusLoading) return
+
+        navigate(hasPassword ? PopupRoutes.Wallet : PopupRoutes.SetPaymentPassword, { replace: true })
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/index.tsx]: Merge it into the previous `useEffect` that contains `navigate` either. And please avoid navigating in `useEffect`. It causes a flash of blank pages.

File: packages/mask/src/extension/popups/pages/Wallet/components/StartUp/index.tsx [hunk 10]
```diff
@@ -1,134 +1,123 @@
...
-        await Services.Helper.removePopupWindow()
-    }, [location.search])
+    const onEnterCreateWallet = useCallback(
+        async (route_: DashboardRoutes) => {
+            const route = hasPassword ? route_ : DashboardRoutes.CreateMaskWalletForm
+            await browser.tabs.create({
+                active: true,
+                url: browser.runtime.getURL(`/dashboard.html#${route}`),
+            })
+            if (navigator.userAgent.includes('Firefox')) {
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/components/StartUp/index.tsx]: Use `Sinffings` stead.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 6 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts]: User has decided to do  this dangerous action.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 8 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts:240]: No longer require password

File: packages/dashboard/src/pages/CreateMaskWallet/context.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+function useContext() {
+    const location = useLocation()
+    const wallets = useWallets()
+
+    const resetWallets = useCallback(async () => {
+        const password = location.state?.password
+        if (location.state?.isReset && wallets.length) {
+            await PluginServices.Wallet.resetPassword(password)
+            for (const wallet of wallets) {
+                await Web3.removeWallet?.(wallet.address, '', {
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/CreateMaskWallet/context.ts]: Use `Web3.removeWallets` stead.

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 8 above)
Comment (guanbinrui) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts:240]: It's **NOT SAFTY** to restore mnemonic words or a private key w/o a password. The Mask SDK required a password for the encryption of those critical information. This means only the user who knows the password has the right to read information from the storage.

File: packages/mask/src/extension/popups/pages/Wallet/SetPaymentPassword/index.tsx [hunk 12]
```diff
@@ -92,64 +149,148 @@ const SetPaymentPassword = memo(() => {
...
+                        </div>
+                    </>
+                )}
+            </Box>
+            {isCreating ? null : (
+                <Box className={classes.navigatorWrapper}>
+                    <Navigator />
+                </Box>
+            )}
+            {isCreating ? (
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/pages/Wallet/SetPaymentPassword/index.tsx]: Why not separate into two components?

File: packages/mask/src/extension/popups/pages/Wallet/SwitchWallet/index.tsx [hunk 13]
```diff
@@ -64,11 +63,10 @@ const SwitchWallet = memo(function SwitchWallet() {
         }
     }, [wallets, history])
     const handleImport = useCallback(() => {
-        if (navigator.userAgent.includes('Firefox')) {
-            Services.Helper.openPopupWindow(PopupRoutes.ImportWallet)
-            return
-        }
-        navigate(PopupRoutes.ImportWallet)
+        browser.tabs.create({
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/SwitchWallet/index.tsx]:
```suggestion
        await browser.tabs.create({
```

File: packages/mask/src/plugins/WalletService/services/wallet/index.ts (hunk 8 above)
Comment (guanbinrui) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts:240]: You can update get password function as:

```ts
export async function INTERNAL_getPasswordRequired() {
    if (await hasPassword()) {
        const password_ = await INTERNAL_getPassword()
        if (!password_) throw new Error('No password set yet or expired.')
        return password_
    } else {
        return 'MASK'
    }
}
```

Comment (guanbinrui) [reply packages/mask/src/plugins/WalletService/services/wallet/index.ts:240]: There is no need for the front-end to pass down the default password.

File: packages/mask/src/plugins/WalletService/services/wallet/database/secret.ts [hunk 14]
```diff
@@ -66,13 +82,44 @@ export async function encryptSecret(password: string) {
...
+/**
+ * Update the user password which is used for encrypting the master secret.
+ * @param oldPassword
+ * @param newPassword
+ */
 export async function updateSecret(oldPassword: string, newPassword: string) {
     const secret = await getSecret()
     if (!secret) throw new Error('Failed to update secret.')
 
+    if (newPassword === getDefaultUserPassword() || !newPassword) throw new Error('Invalid password.')
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/wallet/database/secret.ts]: Why !newPassword?

File: packages/mask/src/plugins/WalletService/services/helpers.ts [hunk 15]
```diff
@@ -4,6 +4,13 @@ import { formatEthereumAddress } from '@masknet/web3-shared-evm'
 import type { IDBPSafeTransaction } from '../../../../background/database/utils/openDB.js'
 import type { LegacyWalletRecordInDatabase } from '../database/types.js'
 
+/**
+ * The default user password for master secret encryption.
+ */
+export function getDefaultUserPassword() {
+    return 'MASK NETWORK'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/WalletService/services/helpers.ts]: Add .reverse()

**Pull request merged**